src/platforms/hosted/bmp_libusb.c

```diff
@@ -47,291 +106,373 @@ void libusb_exit_function(bmp_info_s *info)
+		for (size_t iface = 0; iface < config->bNumInterfaces; ++iface) {
+			const libusb_interface_s *interface = &config->interface[iface];
+			for (int descriptorIndex = 0; descriptorIndex < interface->num_altsetting; ++descriptorIndex) {
+				const libusb_interface_descriptor_s *descriptor = &interface->altsetting[descriptorIndex];
+				uint8_t string_index = descriptor->iInterface;
+				if (string_index == 0)
+					continue;
+				if (libusb_get_string_descriptor_ascii(handle, string_index, (unsigned char *)version, buffer_size) < 0)
+					continue; /* We failed but that's a soft error at this point. */
+
+				if (strstr(version, "Version") != NULL) {
+					break;
+				}
+			}
+		}
 	}
+}
 
-	for (uint8_t i = 0; i < conf->bNumInterfaces; ++i) {
-		const libusb_interface_descriptor_s *interface = &conf->interface[i].altsetting[0];
+bool process_cmsis_interface_probe(
+	libusb_device_descriptor_s *device_descriptor, libusb_device *device, probe_info_s **probe_list, bmp_info_s *info)
+{
+	(void)info;
+	char *serial;
+	char *manufacturer;
+	char *product;
+	char *version;
 
-		if (!interface->iInterface)
-			continue;
+	libusb_device_handle *handle;
+	bool cmsis_dap = false;
 
-		res = libusb_get_string_descriptor_ascii(
-			handle, interface->iInterface, (uint8_t *)interface_string, sizeof(interface_string));
-		if (res < 0) {
-			DEBUG_ERROR("libusb_get_string_descriptor_ascii() failed: %s\n", libusb_strerror(res));
-			continue;
-		}
+	libusb_config_descriptor_s *config;
 
-		if (!strstr(interface_string, "CMSIS"))
-			continue;
-		type = BMP_TYPE_CMSIS_DAP;
+	if (libusb_get_active_config_descriptor(device, &config) == 0 && libusb_open(device, &handle) == 0) {
+		char read_string[128];
 
-		if (interface->bInterfaceClass == 0xffU && interface->bNumEndpoints == 2U) {
-			info->interface_num = interface->bInterfaceNumber;
+		for (size_t iface = 0; iface < config->bNumInterfaces && !cmsis_dap; ++iface) {
+			const libusb_interface_s *interface = &config->interface[iface];
+			for (int descriptorIndex = 0; descriptorIndex < interface->num_altsetting; ++descriptorIndex) {
+				const libusb_interface_descriptor_s *descriptor = &interface->altsetting[0];
+				uint8_t string_index = descriptor->iInterface;
+				if (string_index == 0)
+					continue;
+				if (libusb_get_string_descriptor_ascii(
+						handle, string_index, (unsigned char *)read_string, sizeof(read_string)) < 0)
+					continue; /* We failed but that's a soft error at this point. */
 
-			for (uint8_t j = 0; j < interface->bNumEndpoints; ++j) {
-				const uint8_t n = interface->endpoint[j].bEndpointAddress;
-				if (n & 0x80U)
-					info->in_ep = n;
-				else
-					info->out_ep = n;
+				if (strstr(read_string, "CMSIS") != NULL) {
+					if (device_descriptor->idVendor == VENDOR_ID_ORBCODE &&
+						device_descriptor->idProduct == PRODUCT_ID_ORBTRACE) {
+						orbtrace_read_version(device, handle, read_string, sizeof(read_string));
+						version = strdup(read_string);
+					} else {
```

Comment:
Many of these if and else blocks have extra braces that can be removed.

Reply:
Excess braces removed.